### PR TITLE
unsubscripe app from waypoints when it is unregistered

### LINKED ISSUE #2461: SDL does not  remove app from waypoints list
### Bug Report
SDL does not  remove app from waypoints list 

##### Reproduction Steps
1. In case of expected disconnect SDL never removes the app
from subscribe way points list. As a result during next connection SDL
does not send SubscribeWayPoits request to HMI.

##### Expected Behavior
SDL remove app from waypoints list 

##### Observed Behavior
SDL does not  remove app from waypoints list 

##### OS & Version Information
* OS/Version:
* SDL Core Version: 
* Testing Against:

### LINKED ISSUE #3372: Core Unsubscribes from HMI waypoints when it should not
### Bug Report
Code in `ApplicationManagerImpl::UnregisterApplication` seems to assume that the app being unregistered is subscribed to waypoints, and checks if there is only one app in the `subscribed_way_points_apps_list_ ` before calling `MessageHelper::SendUnsubscribedWayPoints` 

https://github.com/smartdevicelink/sdl_core/blob/0ce76256232bad18bac986e7e68f35c2ae510a29/src/components/application_manager/src/application_manager_impl.cc#L3219

##### Reproduction Steps
1. Connect two apps
2. App1 subscribes to way points
3. App2 disconnects

##### Expected Behavior
Core does not send UnsubscribeWayPoints to HMI

##### Observed Behavior
Core sends UnsubscribeWayPoints to HMI

##### OS & Version Information
* OS/Version: any
* SDL Core Version: 6.1.1 develop
* Testing Against: ATF

##### Test Case, Sample Code, and / or Example App
https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2408

### PR DESCRIPTION
**Pull request by iCollin**
Fixes #2461
Fixes #3372 

This PR is **ready** for review.

### Risk
This PR makes **minor** API changes.

### Testing Plan
[ATF tests to be created](https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2408)

### Summary
if `IsAppIdSubscribedForWayPoints` then `UnsubscribeAppIdFromWayPoints` and also if `1 == subscribed_for_way_points_app_count` then `MessageHelper::SendUnsubscribedWayPoints`

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Comment by jacobkeeler** [comment]
Also fix style

File: src/components/application_manager/include/application_manager/application_manager_impl.h [hunk 1]
```diff
@@ -283,6 +291,12 @@ class ApplicationManagerImpl
    */
   void SubscribeAppForWayPoints(ApplicationSharedPtr app) OVERRIDE;
```
[the remark this answers is not in the document]
Comment (iCollin) [reply src/components/application_manager/include/application_manager/application_manager_impl.h:298]: good idea. done in https://github.com/smartdevicelink/sdl_core/pull/3373/commits/cc68b603ce02be7f9b9852cdd23aee989168dba3

File: src/components/application_manager/include/application_manager/application_manager_impl.h [hunk 2]
```diff
@@ -283,6 +291,12 @@ class ApplicationManagerImpl
...
 
+  /**
+   * @brief Unsubscribe Application for way points
+   * @param Application id
+   */
+  void UnsubscribeAppIdFromWayPoints(uint32_t app_id) OVERRIDE;
```
[the remark this answers is not in the document]
Comment (iCollin) [reply src/components/application_manager/include/application_manager/application_manager_impl.h]: good suggestion, i don't know about deprecating but I went the two signatures route in https://github.com/smartdevicelink/sdl_core/pull/3373/commits/cc68b603ce02be7f9b9852cdd23aee989168dba3

File: src/components/application_manager/include/application_manager/application_manager_impl.h [hunk 3]
```diff
@@ -269,6 +269,14 @@ class ApplicationManagerImpl
    */
   bool IsAppTypeExistsInFullOrLimited(ApplicationConstSharedPtr app) const;
 
+  /**
+   * @brief Checks if Application is subscribed for way points
+   * @param Application id
+   * @return true if Application is subscribed for way points
+   * otherwise false
+   */
+  bool IsAppIdSubscribedForWayPoints(uint32_t app_id) const OVERRIDE;
```
[the remark this answers is not in the document]
Comment (iCollin) [reply src/components/application_manager/include/application_manager/application_manager_impl.h]: good suggestion, i don't know about deprecating but I went the two signatures route in https://github.com/smartdevicelink/sdl_core/pull/3373/commits/cc68b603ce02be7f9b9852cdd23aee989168dba3

File: src/components/application_manager/src/application_manager_impl.cc [hunk 4]
```diff
@@ -3222,9 +3222,12 @@ void ApplicationManagerImpl::UnregisterApplication(
...
   }
-  if (1 == subscribed_for_way_points_app_count) {
-    LOG4CXX_ERROR(logger_, "Send UnsubscribeWayPoints");
-    MessageHelper::SendUnsubscribedWayPoints(*this);
+  if (IsAppIdSubscribedForWayPoints(app_id)) {
+    if (1 == subscribed_for_way_points_app_count) {
+      LOG4CXX_ERROR(logger_, "Send UnsubscribeWayPoints");
+      MessageHelper::SendUnsubscribedWayPoints(*this);
+    }
+    UnsubscribeAppIdFromWayPoints(app_id);
```
[the remark this answers is not in the document]
Comment (iCollin) [reply src/components/application_manager/src/application_manager_impl.cc]: suggestion applied in https://github.com/smartdevicelink/sdl_core/commit/cc68b603ce02be7f9b9852cdd23aee989168dba3

**Pull request merged**